src/socket/mod.rs

```diff
@@ -0,0 +1,311 @@
+fn with_timeout<F, A>(timeout: Timeout, action: F) -> Result<(Timeout, A), ()>
+where
+    F: FnOnce(&Timeout) -> A,
+{
+    // start a timer
+    let i = Instant::now();
+    // run the action with the full timeout
+    let res = action(&timeout);
+    match timeout {
+        Timeout::Infinite => Ok((Timeout::Infinite, res)),
+        Timeout::D(Millis(u)) => match Millis::try_from(i.elapsed()) {
+            // more time elapsed than we can hold in a timeout
+            // so it must have blown the allowed timeout
+            Err(_) => Err(()),
+            Ok(Millis(elapsed)) => match u.checked_sub(elapsed) {
+                // no time remaining
+                None => Err(()),
+                // return a new timeout with remaining time
+                Some(u2) => Ok((Timeout::D(Millis(u2)), res)),
+            },
+        },
+    }
+}
+
+impl<'a, R, W> ReadWriter<'a, R, W> {
+    // TODO: &mut UnixStream
+    /// Create from a unix socket.
+    pub fn new(socket: &'a UnixStream) -> ReadWriter<'a, R, W> {
+        ReadWriter {
+            socket,
+            phantom_r: PhantomData,
+            phantom_w: PhantomData,
+        }
+    }
+
+    /// Send a message to the other side and wait for a reply.
+    ///
+    /// The timeout counts for the whole roundtrip.
+    pub fn communicate(&mut self, timeout: Timeout, mes: &W) -> Result<R, ReadWriteError>
+    where
+        R: serde::de::DeserializeOwned,
+        W: serde::Serialize,
+    {
+        let (timeout, write_res) =
+            with_timeout(timeout, |t| self.write(t, mes)).map_err(|()| WriteError::Timeout)?;
+        write_res?;
+        let e = self.read(&timeout)?;
+        Ok(e)
+    }
+
+    /// Listen for a message from the other side and immediately
+    /// send a reply based on the message.
+    ///
+    /// The timeout counts for the whole roundtrip.
+    pub fn react<F>(&mut self, timeout: Timeout, reaction: F) -> Result<R, ReadWriteError>
+    where
+        R: serde::de::DeserializeOwned,
+        W: serde::Serialize,
+        F: FnOnce(&R) -> W,
+    {
+        let (timeout, read_res) =
+            with_timeout(timeout, |t| self.read(t)).map_err(|()| ReadError::Timeout)?;
+        let read = read_res?;
+        self.write(&timeout, &reaction(&read))?;
+        Ok(read)
+    }
+
+    /// Check if the underlying socket timed out when serializing/deserilizing.
+    fn is_timed_out(e: &bincode::ErrorKind) -> bool {
+        match e {
+            bincode::ErrorKind::Io(io) => match io.kind() {
+                std::io::ErrorKind::TimedOut => true,
+                _ => false,
+            },
+            _ => false,
+        }
+    }
+
+    /// Wait for a message to arrive.
+    pub fn read(&self, timeout: &Timeout) -> Result<R, ReadError>
+    where
+        R: serde::de::DeserializeOwned,
+    {
+        let timeout_socket = timeout::TimeoutReadWriter::new(self.socket, timeout);
+
+        // XXX: “If this returns an Error, `reader` may be in an invalid state”.
```

Comment:
I think that means the socket is tainted and should be restarted, since a read may resume in a weird spot in the data stream.

Reply:
So in that case we’d have to deallocate the unix socket and create a new one?

Reply:
I think so

Reply:
https://github.com/target/lorri/issues/102